components/src/Tabs/TabScrollIndicators.js

```diff
@@ -0,0 +1,155 @@
+import styled from "styled-components";
+import PropTypes from "prop-types";
+import React from "react";
+import smoothscroll from "smoothscroll-polyfill";
+import theme from "../theme";
+import TabScrollIndicator from "./TabScrollIndicator";
+
+smoothscroll.polyfill();
+
+const TabScrollIndicatorContainer = styled.div(({ width }) => ({
+  position: "absolute",
+  width,
+  height: theme.space.x5
+}));
+
+class TabScrollIndicators extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      scrollLeft: 0
+    };
+
+    this.handleIndicatorClick = this.handleIndicatorClick.bind(this);
+    this.setScrollLeftState = this.setScrollLeftState.bind(this);
+    this.getScrollLeftValueByTabIndex = this.getScrollLeftValueByTabIndex.bind(this);
+    this.contentHiddenLeft = this.contentHiddenLeft.bind(this);
+    this.contentHiddenRight = this.contentHiddenRight.bind(this);
+    this.handleScroll = this.handleScroll.bind(this);
+  }
+
+  setScrollLeftState(scrollLeft) {
+    this.setState({ scrollLeft }, this.applyScrollLeft);
+  }
+
+  getScrollLeftValueByTabIndex(index) {
+    const { tabRefs } = this.props;
+    let scrollLeftSum = 0;
+    for (let i = 0; i < index; i += 1) {
+      scrollLeftSum += tabRefs[i].offsetWidth;
+    }
+    return scrollLeftSum;
+  }
+
+  contentHiddenRight() {
+    const { tabContainerRef } = this.props;
+    if (!tabContainerRef.current) {
+      return false;
+    }
+    return (
+      tabContainerRef.current.scrollLeft + tabContainerRef.current.offsetWidth < tabContainerRef.current.scrollWidth
+    );
+  }
+
+  contentHiddenLeft() {
+    const { tabContainerRef } = this.props;
+    if (!tabContainerRef.current) {
+      return false;
+    }
+    return (
+      tabContainerRef.current.scrollLeft !== 0 &&
+      tabContainerRef.current.offsetWidth < tabContainerRef.current.scrollWidth
+    );
+  }
+
+  findLastVisibleTab() {
+    const { tabContainerRef, tabRefs, indicatorWidth } = this.props;
+    const rightMarker = tabContainerRef.current.scrollLeft + tabContainerRef.current.offsetWidth - indicatorWidth;
+    let scrollLeftSum = 0;
+
+    for (let i = 0; i < tabRefs.length; i += 1) {
+      scrollLeftSum += tabRefs[i].offsetWidth;
+      if (rightMarker <= scrollLeftSum) {
+        return i;
+      }
+    }
+    return null;
+  }
+
+  findFirstVisibleTab() {
+    const { tabContainerRef, tabRefs, indicatorWidth } = this.props;
+    const leftMarker = tabContainerRef.current.scrollLeft + indicatorWidth;
+    let scrollLeftSum = 0;
+
+    for (let i = 0; i < tabRefs.length; i += 1) {
+      scrollLeftSum += tabRefs[i].offsetWidth;
+      if (leftMarker <= scrollLeftSum) {
+        return i;
+      }
+    }
+    return null;
+  }
+
+  handleScroll() {
```

Comment:
Does this trigger when the scrolling ends? Or several times during a scroll?

Reply:
Instead of going with a debounce or a throttle (as it would cause the ScrollIndicator on each side not appear/disappear when hitting the end or start of the container) I went with the approach of removing the scrollLeft state altogether. Instead I added contentHiddenLeft and contentHiddenRight boolean states which update only when they change which causes the ScrollIndicators to re-render. onScroll will now only have to run the if statement which is much less taxing than constantly setting state causing re-renders